### PR TITLE
[WIP] Typescript conversion

### PR DESCRIPTION
**Pull request by l2succes**
This migrates the codebase from flow to typescript for type definitions.
Fixed just enough to get top level containers to run in the simulator. 

_You may need to run `tsc --watch` if files are not automatically being compiled by VS Code (didn't work for me but will fix)._

To-do list:

- [ ] Add types to remaining classes
- [x] Fix tests
- [ ] Fix hot reloading

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>No CHANGELOG added.</td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by l2succes** [comment]
@sarahscott I'm currently working on defining types and fixing errors for the `Gene` container and its sub-components. So feel free to work on any of the other containers. I should have that work pushed to this branch hopefully by EOD so you can be aligned on conventions.

**Comment by l2succes** [comment]
@orta Sounds good, I added the hack for now but hot-reloading doesn't work yet so either I'll try to fix it or switch to using your fork.

**Comment by sarahscott** [comment]
@l2succes fantastic! I'll start on the Artist container/components.

File: src/lib/components/tab_view.tsx [hunk 1]
```diff
@@ -1,14 +1,27 @@
...
+export default class TabView extends React.Component<Props, {}> {
+  static propTypes = {
+    ...SwitchView.propTypes
+  }
+
   render() {
     const { children, ...props } = this.props
     return (
       <View>
-        <SwitchView style={styles.switch} {...props} />
```
Comment (sarahscott) [inline src/lib/components/tab_view.tsx]: This spread operator gave me a lot of trouble - I did appreciate the elegance of this class though so if someone knows how to pass the props in this way without ts errors, it'll be very appreciated 😄 .

File: src/lib/components/artist/articles/article.tsx [hunk 2]
```diff
@@ -1,14 +1,24 @@
...
+      name
+    }
+    thumbnail_image: {
+      url
+    }
+    thumbnail_title: string
+  }
+}
+
+class Article extends React.Component<Props, {}> {
```
Comment (alloy) [inline src/lib/components/artist/articles/article.tsx]: I think the second `state` type should be `null` if there’s no state.

File: tsconfig.json [hunk 3]
```diff
@@ -1,11 +1,27 @@
...
-        "allowJs": true
+        "jsx": "react",
+        "target": "ESNext",
+        "module": "es2015",
+        "rootDir": "src",
+        "outDir": "build",
+        "allowJs": true,
+        "sourceMap": true,
+        "moduleResolution": "node",
+        "strictNullChecks": true
```
Comment (orta) [inline tsconfig.json]: this is the good stuff

File: tsconfig.json [hunk 4]
```diff
@@ -1,11 +1,27 @@
...
         "node_modules",
         "Pod/Assets",
         "Example/Build",
-        "externals/"
-    ]
+        "externals",
+        "build",
+        "index.ios.js"
+    ],
+    "compileOnSave": true
```
Comment (orta) [inline tsconfig.json]: this we want to eliminate via https://github.com/orta/systems-theory/issues/1 and https://github.com/facebook/react-native/issues/11932

File: Example/EmissionTests/TestApps.js [hunk 5]
```diff
@@ -1,24 +1,21 @@
-/* @flow */
-'use strict';
-
-import Emission from 'emission';
-import React from 'react';
-import ReactNative from 'react-native';
-const { View } = ReactNative;
-const { TestModule } = ReactNative.NativeModules;
+import Emission from 'emission'
+import React from 'react'
```
Comment (alloy) [inline Example/EmissionTests/TestApps.js]: Shouldn’t this be?

```
import * as React from 'react'
```

Comment (l2succes) [inline Example/EmissionTests/TestApps.js]: Yeah that's correct, still working on converting all the tests. I'll have it updated soon

Comment (orta) [inline Example/EmissionTests/TestApps.js]: nah, just checked, this doesn't need to change - it's still a JS file, and should stay one probably.  It's only referenced by the iOS Sim Test Runner.

File: src/lib/components/load_failure_view.tsx [hunk 6]
```diff
@@ -1,19 +1,17 @@
...
 import {
   requireNativeComponent,
 } from 'react-native'
 
-export default class LoadFailureView extends React.Component {
+export default class LoadFailureView extends React.Component<any,any> {
+  static propTypes = {
+    // A callback that is called when the user requests a retry.
+    onRetry: React.PropTypes.func,
+  }
```
Comment (alloy) [inline src/lib/components/load_failure_view.tsx]: You should move this to a Props interface instead.

Comment (l2succes) [inline src/lib/components/load_failure_view.tsx]: We need to have propTypes defined for native components but yes this should be defined in the Props interface as well

Comment (alloy) [inline src/lib/components/load_failure_view.tsx]: Ah right, forgot about that 👍 

This should probably be documented in the README:

* JS components: only a TS interface (which inherits from x and y)
* Native components: also add prop types

**Comment by l2succes** [comment]
Closing this PR in favour of the following one https://github.com/artsy/emission/pull/453

**Pull request closed without merging**